pymc3/distributions/distribution.py

```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+
+        # Raise on unsupported parametrization
+        if shape is not None and dims is not None:
+            raise ValueError("Passing both `shape` ({shape}) and `dims` ({dims}) is not supported!")
+        if dims is not None and size is not None:
+            raise ValueError("Passing both `dims` ({dims}) and `size` ({size}) is not supported!")
+        if shape is not None and size is not None:
+            raise ValueError("Passing both `shape` ({shape}) and `size` ({size}) is not supported!")
+
+        # Warn about discouraged parametrization
+        if shape is not None and not isinstance(shape, (list, tuple)):
+            warnings.warn("The `shape` parameter should be a list or tuple.", UserWarning)
+            shape = (shape,)
+        if dims is not None and not isinstance(dims, (list, tuple)):
+            warnings.warn("The `dims` parameter should be a list or tuple.", UserWarning)
+            dims = (dims,)
+        if size is not None and not isinstance(size, (list, tuple)):
+            warnings.warn("The `size` parameter should be a list or tuple.", UserWarning)
+            size = (size,)
+
+        if size is not None and Ellipsis in size:
+            raise ValueError("The `size` parameter cannot contain an Ellipsis. Actual: {size}")
+        if not _valid_ellipsis_position(shape):
+            raise ValueError(
+                f"Ellipsis in `shape` may only appear in the last position. Actual: {shape}"
+            )
+        if not _valid_ellipsis_position(dims):
+            raise ValueError(
+                f"Ellipsis in `dims` may only appear in the last position. Actual: {dims}"
+            )
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        # Create the RV without specifying size or testval.
+        # The size will be expanded later (if necessary) and only then the testval fits.
+        rv_native = cls.dist(*args, rng=rng, testval=None, size=None, **kwargs)
+        implied_ndim = rv_native.ndim
+        implied_dims = ...  # TODO: infer dimension names from Variables!
+
+        if shape is None and dims is None and size is None:
+            size = ()
+        elif size is not None:
+            # User already specified how to expand the RV
+            pass
+        elif shape is not None:
+            # Infer size from shape
+            if Ellipsis in shape:
+                size = tuple(shape[:-1])
+            else:
+                size = tuple(shape[: len(shape) - implied_ndim])
+                if size:
+                    warnings.warn(
+                        f"The specified shape {shape} has {len(shape)} dimensions. "
+                        f"This is more than the {implied_ndim} dimensions implied by the distribution parameters. "
+                        f"To replicate the RV beyond its implied dimensionality use `size={size}` instead.",
+                        UserWarning,
+                    )
```

Comment:
Does this mean that doing `pm.Normal("a", 0, 1, shape=(3, 20))` will raise a user warning? If so, can we deprecate and/or remove `shape`? 

I have been lost for a while, I think discussing this on the meeting confused me more than it helped. I though we would only keep one of `size`/`shape`, seeing this I think shape should be removed, but I have been partial to `dims`+`size` since the beginning.

Reply:
Yes, your example will raise this warning.
And it's correct that with the current state of this PR the full shape specification does not add anything. My first priority was to get things working, now the next step is, for example, to add the automatic shape assert whenever `shape` is passed.

Reply:
That's not what we discussed. Most of us were in agreement that explicit shapes are better than implicit size.

Reply:
I do not recall _most_ people agreeing on anything involving this entire matter.

The point in favor of the `shape` parameter has only ever been that it's somehow more "explicit" and/or better than a comment stating the expected shape, and this point somehow outweighs the effort/know-how required to specify it correctly and the unnecessary limitations it imposes.

Remember, we could accomplish the same thing without all this extra project-level code if we simply wrote a small blurb in the documentation describing how people can use Aesara's `SpecifyShape` instead.

For example, if one wants to be explicit about a random variable's shape:
```python
import numpy as np

import aesara
import aesara.tensor as at

from aesara.tensor.shape import specify_shape

import pymc3 as pm


mu = at.vector("mu")
Sigma = at.matrix("Sigma")

X = pm.MvNormal.dist(mu, Sigma, size=(2, 3))

# This is it; one extra line that accomplishes the exact same thing as the old
# `shape` parameter
X = specify_shape(X, (2, 3, 4))
```

What does this do for us?  Not much, but here's an example:
```python
>>> X.eval({mu: np.zeros(4), Sigma: np.eye(4)})
array([[[ 0.84616091, -1.38344376, -1.22742971, -0.66795514],
        [ 0.90845087,  0.366118  ,  0.29222136, -0.75515633],
        [ 0.15882859, -0.97699321, -1.09552517, -1.24519882]],

       [[-1.05805731, -2.63053756, -0.30468915,  1.21047848],
        [-1.10110766, -0.93431473, -0.00385619, -0.97005804],
        [ 1.07159636, -1.02907491, -0.39692291,  0.39597758]]])

>>> X.eval({mu: np.zeros(2), Sigma: np.eye(2)})
...
AssertionError: SpecifyShape: dim 2 of input has shape 2, expected 4.
...
```

Unfortunately, this triviality is taking precious developer time and preventing us from actually moving forward (e.g. merging into `master`), so we need consider whether or not this `shape` thing is really more important than any or all of the implemented, in-progress, and planned `v4` improvements.